lib/LANraragi/Controller/Plugins.pm

```diff
@@ -148,7 +148,30 @@ sub process_upload {
     #Check if this is a Perl package ("xx.pm")
     if ( $filename =~ /^.+\.(?:pm)$/ ) {
 
-        my $dir = getcwd() . ("/lib/LANraragi/Plugin/");
+        #Check plugin type
+        my $filetext   = $file->slurp;
+        my $plugintype = "";
+
+        if ($filetext =~ /package LANraragi::Plugin::(Login|Metadata|Scripts)::/) {
+            $plugintype = $1;
+        } else {
+            $logger->error("Could not find a valid plugin package type in the plugin "
+              . "\"$filename\"!");
+
+            $self->render(
+                json => {
+                    operation => "upload_plugin",
+                    name      => $file->filename,
+                    success   => 0,
+                    error     => "Could not find a valid plugin package type in the "
+                      . "plugin \"$filename\"!"
+                }
+            );
```

Comment:
Put the error message in a single variable to avoid retyping it:  
```suggestion
            my $errormess= "Could not find a valid plugin package type in the plugin \"$filename\"!";
            $logger->error($errormess);

            $self->render(
                json => {
                    operation => "upload_plugin",
                    name      => $file->filename,
                    success   => 0,
                    error     => $errormess
                }
            );
```

Reply:
I know. This is just how the existing code does it and, as I mentioned, I tried to stay consistent.

Reply:
Well, the only other place in that file where we spit a success 0 json with an error code has slightly different text between the log and output versions. (Since the output has html tags in it as a quick 'n dirty way to make it fit in the notification) 

Here tho they're exactly the same, so I'll take legibility over consistency. 😛 Thanks for the change.